File: 2015/Day25/main.dart
import 'dart:io';

void main() async {
  final line = await File('input.txt').readAsString();

  print('Result: ${parse(line.trim())}');
}

const int divider = 33554393;
const int initialCode = 20151125;
const int multiplier = 252533;

int getCodeAtPosition(int row, int col) {
  final diagonal = row + col - 1;
  final sequenceNumber = (diagonal * (diagonal - 1)) ~/ 2 + col;

  int code = initialCode;
  for (int i = 1; i < sequenceNumber; ++i) {
    code = (code * multiplier) % divider;
  }

  return code;
}

int parse(String line) {
  final regex = RegExp(r'Enter the code at row (\d+), column (\d+).');
  final match = regex.firstMatch(line)!;
  final row = int.parse(match.group(1)!);
  final col = int.parse(match.group(2)!);

  return getCodeAtPosition(row, col);
}

File: 2015/Day25/main.lua
local divider = 33554393
local initial_code = 20151125
local multiplier = 252533

local file = io.open('input.txt', 'r')
if not file then return nil end
local row, col = file:read('*all'):match('Enter the code at row (%d+), column (%d+).')
file:close()

row, col = tonumber(row), tonumber(col)

local diagonal = row + col - 1
local sequence_number = (diagonal * (diagonal - 1)) // 2 + col

local code = initial_code
for i = 2, sequence_number do
    code = (code * multiplier) % divider
end

print('Result: ' .. code)
